src/ray/raylet/scheduling/cluster_task_manager.cc

```diff
@@ -295,6 +315,24 @@ void ClusterTaskManager::DispatchScheduledTasksToWorkers(
         continue;
       }
 
+      if (sched_cls_info.running_tasks.size() >= sched_cls_info.capacity) {
+        if (get_time_() > sched_cls_info.next_update_time) {
+          // Don't increase the capacity on the very first update since that
+          // will unblock a new task immediately, instead of after the timeout.
+          if (sched_cls_info.num_updates > 0) {
+            sched_cls_info.capacity++;
+          }
+
+          double wait_time =
+              (1e6 * sched_cls_cap_interval_ms_) * (1L << sched_cls_info.num_updates++);
+          sched_cls_info.next_update_time = get_time_() + wait_time;
+          sched_cls_info.timer =
+              execute_after_([this]() { ScheduleAndDispatchTasks(); }, wait_time);
+        }
+        work_it++;
+        continue;
```

Comment:
Shouldn't we break if we are at capacity and the timer is still active? Because everything else in this class should also be blocked.

Reply:
Updated to make it a little more precise, but the capacity may have been increased here.